src/models/converters/transfers.rs

```diff
@@ -170,24 +130,39 @@ impl EtherTransferDto {
     }
 }
 
-fn token_info_with_fallback<T>(
-    info_provider: &mut dyn InfoProvider,
-    token_address: &String,
-    token_info: Option<T>,
-    expected_type: TokenType,
-    fallback_mapper: impl Fn(&TokenInfo) -> T,
-) -> Option<T> {
-    if token_info.is_some() {
-        return token_info;
-    }
-    match info_provider.token_info(token_address) {
-        Ok(token) => {
-            if token.token_type == expected_type {
-                Some(fallback_mapper(&token))
-            } else {
-                None
-            }
-        }
-        Err(_) => None,
+fn build_transfer_info(
+    token_info: Option<&TokenInfo>,
+    default_token_type: TokenType,
+    token_address: &str,
+    element: &str,
+) -> TransferInfo {
+    match token_info
+        .map(|it| it.token_type.to_owned())
+        .unwrap_or(default_token_type)
+    {
+        TokenType::Erc20 => TransferInfo::Erc20(Erc20Transfer {
+            token_address: token_address.to_owned(),
+            token_name: token_info.map(|it| it.name.to_owned()),
+            token_symbol: token_info.map(|it| it.symbol.to_owned()),
+            logo_uri: token_info.map(|it| it.logo_uri.to_owned()).flatten(),
+            decimals: token_info.map(|it| it.decimals.to_owned()),
+            value: element.to_owned(),
+        }),
+        TokenType::Erc721 => TransferInfo::Erc721(Erc721Transfer {
+            token_address: token_address.to_owned(),
+            token_id: element.to_owned(),
+            token_name: token_info.map(|it| it.name.to_owned()),
+            token_symbol: token_info.map(|it| it.symbol.to_owned()),
+            logo_uri: token_info.map(|it| it.logo_uri.to_owned()).flatten(),
```

Comment:
Why do we have to flatten this?

Reply:
Because here we have `Option<TokenInfo>` and within `token_info` we have `Option<String>` for `logo_uri`, so we are in a "option of an option" situation, with flatten you get rid of all the nested options.